torch/_tensor_str.py

```diff
@@ -277,31 +277,34 @@ def _vector_str(self):
         ident = ' '
     if self.numel() < PRINT_OPTS.threshold:
         return (strt +
-                '\n'.join(ident + fmt.format(val / scale) for val in self) +
+                '\n'.join(ident + fmt.format(val.item() / scale) for val in self) +
                 '\n')
     else:
         return (strt +
-                '\n'.join(ident + fmt.format(val / scale) for val in self[:n]) +
+                '\n'.join(ident + fmt.format(val.item() / scale) for val in self[:n]) +
                 '\n' + (ident + dotfmt.format(u"\u22EE")) +
-                '\n'.join(ident + fmt.format(val / scale) for val in self[-n:]) +
+                '\n'.join(ident + fmt.format(val.item() / scale) for val in self[-n:]) +
                 '\n')
 
 
 def _str(self, include_footer=True):
-    if self.ndimension() == 0:
+    empty = self.numel() == 0
+    dim = self.dim()
+
+    if empty:
         strt = ''
-    elif self.ndimension() == 1:
+    elif dim == 0:
+        strt = _vector_str(self.unsqueeze(0))
+    elif dim == 1:
         strt = _vector_str(self)
-    elif self.ndimension() == 2:
+    elif dim == 2:
         strt = _matrix_str(self)
     else:
         strt = _tensor_str(self)
 
     if include_footer:
-        size_str = 'x'.join(str(size) for size in self.size())
-        size_str_prefix = 'of size ' if self.ndimension() > 0 else 'with no dimension'
+        size_str = '({})'.format(','.join(str(size) for size in self.shape))
```

Comment:
Use what's in variable.py:
size_str = '(' + ','.join(str(size) for size in tensor.size()) + (',)' if len(tensor.size()) == 1 else ')')
(this matches numpy behavior in that it prints (1,) instead of (1)

Also does this handle sparse correctly?  It's probably easier to just pull out the include_footer stuff and copy the code from variable.py into here.

Reply:
oh you deleted what was in variable.py; just move it instead.

Reply:
It also prints (,) which is wrong

Reply:
Why don't we just use `str(tuple(self.shape))`?

Reply:
It doesn't print (,).

Reply:
I'm fine with str(tuple(self.shape)); the suggested code doesn't have spaces between elements and it seemed nicer to be more compact but I don't have a strong preference.